pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -90,6 +92,136 @@ pub struct InMemoryLayerInner {
     resource_units: GlobalResourceUnits,
 }
 
+/// Support the same max blob length as blob_io, because ultimately
+/// all the InMemoryLayer contents end up being written into a delta layer,
+/// using the [`crate::tenant::blob_io`].
+const MAX_SUPPORTED_BLOB_LEN: usize = crate::tenant::blob_io::MAX_SUPPORTED_BLOB_LEN;
+const MAX_SUPPORTED_BLOB_LEN_BITS: usize = {
+    let trailing_ones = MAX_SUPPORTED_BLOB_LEN.trailing_ones() as usize;
+    let leading_zeroes = MAX_SUPPORTED_BLOB_LEN.leading_zeros() as usize;
+    assert!(trailing_ones + leading_zeroes == std::mem::size_of::<usize>() * 8);
+    trailing_ones
+};
+
+/// See [`InMemoryLayerInner::index`].
+///
+/// For space-efficiency, this value is a bitfield.
+///
+/// Layout:
+/// - 1 bit: `will_init`
+/// - [`MAX_SUPPORTED_BLOB_LEN_BITS`]: `len`
+/// - [`MAX_SUPPORTED_POS_BITS`]: `pos`
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct InMemoryLayerIndexValue(u64);
+
+impl InMemoryLayerIndexValue {
+    /// Derive remaining space for pos.
+    /// TODO: define and enforce a hard limit at the [`crate::tenant::Timeline::should_roll`] level.
+    /// => see also [`Self::does_timeline_should_roll_prevent_failure`]
+    const MAX_SUPPORTED_POS_BITS: usize = {
+        let remainder = 64 - 1 - MAX_SUPPORTED_BLOB_LEN_BITS;
+        if remainder < 32 {
+            panic!("pos can be u32 as per type system, support that");
+        }
+        remainder
+    };
+    const MAX_SUPPORTED_POS: usize = (1 << Self::MAX_SUPPORTED_POS_BITS) - 1;
+
+    // Layout
+    const WILL_INIT_RANGE: Range<usize> = 0..1;
+    const LEN_RANGE: Range<usize> =
+        Self::WILL_INIT_RANGE.end..Self::WILL_INIT_RANGE.end + MAX_SUPPORTED_BLOB_LEN_BITS;
+    const POS_RANGE: Range<usize> =
+        Self::LEN_RANGE.end..Self::LEN_RANGE.end + Self::MAX_SUPPORTED_POS_BITS;
+    const _ASSERT: () = {
+        if Self::POS_RANGE.end != 64 {
+            panic!("we don't want undefined bits for our own sanity")
+        }
+    };
+
+    /// Call this with the checkpoint distance enforced by Timeline::should_roll to check whether
+    /// [`Self`] can accomodate large enough values.
+    ///
+    /// TODO: this check should happen much earlier, ideally at the type system level.
+    /// When cleaning this up, also look into the s3 max file size check that is performed in delta layer writer.
+    /// See also [`Self::MAX_SUPPORTED_POS_BITS`].
+    pub(crate) fn does_timeline_should_roll_prevent_failure(
```

Comment:
nit: This name confuses me. Something like `validate_checkpoint_distance` would make more sense to me. It's also in a strange place. I'd maybe drop it with other tenent config stuff.